src/coreclr/vm/excep.cpp

```diff
@@ -3226,196 +2858,326 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
+    }
+    else if (!pCf->HasFaulted() && stackTraceElem.ip != 0)
     {
-        StackTraceElement* pStackTraceElem;
+        stackTraceElem.ip -= STACKWALK_CONTROLPC_ADJUST_OFFSET;
+        stackTraceElem.flags |= STEF_IP_ADJUSTED;
+    }
 
-        // If we get in here, we'd better have a stack trace array.
-        CONSISTENCY_CHECK(m_pStackTrace != NULL);
+#ifndef TARGET_UNIX // Watson is supported on Windows only
+    SetupWatsonBucket(currentIP, pCf);
+#endif // !TARGET_UNIX
 
-        pStackTraceElem = &(m_pStackTrace[m_dFrameCount]);
+    EX_TRY
+    {
+        struct
+        {
+            StackTraceArray stackTrace;
+            PTRARRAYREF pKeepaliveArray = NULL; // Object array of Managed Resolvers / Loader Allocators of methods that can be collected
+            OBJECTREF keepaliveObject = NULL;
+        } gc;
+
+        GCPROTECT_BEGIN_THREAD(pThread, gc);
 
-        pStackTraceElem->pFunc = pFunc;
+        // Fetch the stacktrace and the keepalive array from the exception object. It returns clones of those arrays in case the
+        // stack trace was created by a different thread.
+        bool wasCreatedByForeignThread = ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->GetStackTrace(gc.stackTrace, &gc.pKeepaliveArray);
 
-        pStackTraceElem->ip = currentIP;
-        pStackTraceElem->sp = currentSP;
+        // The stack trace returned by the GetStackTrace has to be created by the current thread or be NULL.
+        _ASSERTE((gc.stackTrace.Get() == NULL) || (gc.stackTrace.GetObjectThread() == pThread));
 
-        // When we are building stack trace as we encounter managed frames during exception dispatch,
-        // then none of those frames represent a stack trace from a foreign exception (as they represent
-        // the current exception). Hence, set the corresponding flag to FALSE.
-        pStackTraceElem->flags = 0;
+        EnsureStackTraceArray(&gc.stackTrace, gc.stackTrace.Size() + 1);
 
-        // This is a workaround to fix the generation of stack traces from exception objects so that
-        // they point to the line that actually generated the exception instead of the line
-        // following.
-        if (pCf->IsIPadjusted())
+        if (wasCreatedByForeignThread)
         {
-            pStackTraceElem->flags |= STEF_IP_ADJUSTED;
+            // If the stack trace was created by a foreign thread, we need to update the cached keepalive items count.
+            m_keepaliveItemsCount = GetKeepaliveItemsCount(&gc.stackTrace);
         }
-        else if (!pCf->HasFaulted() && pStackTraceElem->ip != 0)
+        else
         {
-            pStackTraceElem->ip -= STACKWALK_CONTROLPC_ADJUST_OFFSET;
-            pStackTraceElem->flags |= STEF_IP_ADJUSTED;
+            if (m_keepaliveItemsCount == -1)
+            {
+                // The m_keepaliveItemsCount was not initialized yet, so we need to calculate it.
+                m_keepaliveItemsCount = GetKeepaliveItemsCount(&gc.stackTrace);
```

Comment:
Should we remember the keep alive item count in the stack trace header, so that we do not have to recompute it here?

Reply:
Done